### PR TITLE
[CompilerPerf] Extends #5112 ideas to Comparers (i.e >, >=, <, <= and LanguagePrimitives.FastGenericComparer)

### PR DESCRIPTION
**Pull request by manofstick**
*This is a standalone improvement, as is #5112, but this is branched off the tail of that PR as I see it being approved and integrated first, as this is still currently a WIP.*

### CODE REVIEW

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 1]
```diff
@@ -744,7 +744,101 @@ namespace Microsoft.FSharp.Core
...
+            let GenericHashObjArray   (iec:System.Collections.IEqualityComparer) (x:obj[]) = ArrayHashing get 0 iec.GetHashCode x
+            let GenericHashByteArray  (x:byte[])  = ArrayHashing get 0 intOfByte x
+            let GenericHashInt32Array (x:int32[]) = ArrayHashing get 0 (fun x -> x) x
+            let GenericHashInt64Array (x:int64[]) = ArrayHashing get 0 int32 x
+
+            // special case - arrays do not by default have a decent structural hashing function
+            let GenericHashArbArray (iec : System.Collections.IEqualityComparer) (x: System.Array) : int =
+                match x.Rank with 
+                | 1 -> ArrayHashing (fun a i -> a.GetValue i) (x.GetLowerBound 0) iec.GetHashCode x
+                | _ -> HashCombine 10 (x.GetLength(0)) (x.GetLength(1))
```
Comment (vasily-kirichenko) [inline src/fsharp/FSharp.Core/prim-types.fs]: why not `if...then...else`?

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 2]
```diff
@@ -744,7 +744,101 @@ namespace Microsoft.FSharp.Core
...
+            // Arrays are structurally hashed through a separate technique.
+            //
+            // "iec" is either fsEqualityComparerUnlimitedHashingER, fsEqualityComparerUnlimitedHashingPER or a CountLimitedHasherPER.
+            let rec GenericHashParamObj (iec : System.Collections.IEqualityComparer) (x: obj) : int =
+                  match x with 
+                  | null -> 0 
+                  | (:? System.Array as a) -> 
+                      match a with 
+                      | :? (obj[]) as oa -> GenericHashObjArray iec oa 
+                      | :? (byte[]) as ba -> GenericHashByteArray ba
```
Comment (vasily-kirichenko) [inline src/fsharp/FSharp.Core/prim-types.fs]: `sbyte[]` will match on this branch as well. Are you sure `GenericHashByteArray` will produce correct result in this case?

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 3]
```diff
@@ -1476,41 +1657,187 @@ namespace Microsoft.FSharp.Core
...
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`3"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`4"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`5"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`6"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`7"
+                                            || isTypeAndGenericArgumentsOK "System.ValueTuple`8"
+                                            || isTypeAndGenericArgumentsOK "Microsoft.FSharp.Collections.FSharpList`1"
+                                            || isTypeAndGenericArgumentsOK "Microsoft.FSharp.Core.FSharpOption`1"
+                                            || isTypeAndGenericArgumentsOK "Microsoft.FSharp.Core.FSharpValueOption`1"
+                                            || isTypeAndGenericArgumentsOK "Microsoft.FSharp.Core.FSharpResult`2"
```
Comment (vasily-kirichenko) [inline src/fsharp/FSharp.Core/prim-types.fs]: This whole condition is fully duplicated on line 1152. Maybe extract it into a function?

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 1 above)
Comment (manofstick) [reply src/fsharp/FSharp.Core/prim-types.fs]: that's what was there previously...

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 2 above)
Comment (manofstick) [reply src/fsharp/FSharp.Core/prim-types.fs]: yep, I know, still WIP...

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 3 above)
Comment (manofstick) [reply src/fsharp/FSharp.Core/prim-types.fs]: not the same - strings

Comment (manofstick) [reply src/fsharp/FSharp.Core/prim-types.fs]: I wouldn't bother too much checking this yet. I'm still a while off completing it...

Comment (manofstick) [reply src/fsharp/FSharp.Core/prim-types.fs]: this is done in the final PR #5307 btw, so assuming the whole chain of PRs is accepted...

**Comment by vasily-kirichenko** [comment]
@manofstick merge conflict!

**Comment by manofstick** [comment]
@vasily-kirichenko

Not being the master of git (or github) I don't know the process to fix.

The conflicted code can just be deleted (it was a line out added, which is now redundant). So the solution is to just ignore the mainline. 

Should I could rebase off master?

**Comment by manofstick** [comment]
Slightly modified tests from #5112... (so they don't necessarily make sense, more just to try as many cases...)

#### 1) https://gist.github.com/manofstick/33fefb8ac6b60d401375be738b521ac3

64-bit

test | original | modified | percent
-----|----------|----------|--------
custom dynamic         | 3599.00 | 1193.17 | 33%  
custom structural      | 3614.00 | 1380.33 | 38%
custom default         | 1190.83 | 1197.33 | 101%
value dynamic          | 2958.33 | 2187.17 | 74%
value structural       | 1248.50 | 1240.00 | 99%
value default          | 1245.33 | 1249.00 | 100%
gen value dynamic      | 6026.33 | 2466.50 | 41%
gen value structural   | 4130.17 | 1503.33 | 36%
gen value default      | 4085.83 | 1524.83 | 37%
ref dynamic            | 2680.83 | 1749.67 | 65%
ref structural         | 1214.00 | 1262.83 | 104%
ref default            | 1314.67 | 1361.67 | 104%
gen ref dynamic        | 5882.17 | 2026.17 | 34%
gen ref structural     | 4343.50 | 1572.33 | 36%
gen ref default        | 4502.50 | 1673.50 | 37%
tuple dynamic          | 5618.67 | 4430.00 | 79%
tuple structural       | 1099.67 | 1070.33 | 97%
tuple default          | 7953.00 | 7844.33 | 99%
value tuple dynamic    | 5719.00 | 1570.17 | 27%
value tuple structural | 5835.83 | 1558.83 | 27%
value tuple default    | 1686.67 | 1535.17 | 91%

32-bit

test | original | modified | percent
-----|----------|----------|--------
custom dynamic         | 11149.33 | 1192.33 | 11% 
custom structural      | 11198.00 | 3798.5 | 34%
custom default         | 1280.50 | 1208.83 | 94%
value dynamic          | 9668.33 | 3648.83 | 38%
value structural       | 1337.00 | 1251.33 | 94%
value default          | 1321.83 | 1259.83 | 95%
gen value dynamic      | 19472.33 | 6444.83 | 33%
gen value structural   | 11084.00 | 3522.83 | 32%
gen value default      | 10975.00 | 3560.67 | 32%
ref dynamic            | 10149.17 | 3766.83 | 37%
ref structural         | 1353.50 | 1281.67 | 95%
ref default            | 1505.67 | 1438.83 | 96%
gen ref dynamic        | 20249.33 | 6484.83 | 32%
gen ref structural     | 12903.83 | 4804.17 | 37%
gen ref default        | 11318.67 | 3961.5 | 35%
tuple dynamic          | 18689.67 | 11176.33 | 60%
tuple structural       | 1326.33 | 1256.67 | 95%
tuple default          | 9330.17 | 8640.83 | 93%
value tuple dynamic    | 17674.67 | 1348 | 8%
value tuple structural | 17620.83 | 1336.83 | 8%
value tuple default    | 1439.33 | 1349.5 | 94%

**Comment by manofstick** [comment]
#### 2) https://gist.github.com/manofstick/04ba0c70c398bf5edeaba9b17d0b17c5

test | original | modified | percent
-----|----------|----------|--------
mapTest 32-bit  | 18169 | 3356 | 18% 
mapTest 64-bit  | 7414 | 4441 | 60%

**Comment by manofstick** [comment]
#### 3) https://gist.github.com/manofstick/bd374436a6d40218b2d836db34c494ca

test | original | modified | percent
-----|----------|----------|--------
32-bit  | 77167 | 68278 | 89% 
64-bit  | 142949 | 74615 | 52%

**Comment by manofstick** [comment]
#### 4) https://gist.github.com/manofstick/258edf7e7d39a76e9d1cdf0316ae2b47

test | original | modified | percent
-----|----------|----------|--------
non-generic 32-bit  | 56360  | 24545 | 44% 
generic 32-bit      | 117833  | 46421 | 39% 
non-generic 64-bit  | 23651 | 18227 | 77%
generic 64-bit      | 44738 | 22051 | 49%

**Comment by manofstick** [comment]
@dsyme - I think this is done.

**Comment by KevinRansom** [comment]
@dsyme , @manofstick , @cartermp -- what do we wnt to do with this PR?  It looks beneficial, but the momentum just died on it!

Should I close it, or will we revisit it in the near future?

Thanks

Kevin

**Comment by KevinRansom** [comment]
@manofstick you will still do these really cool PR's in the future right, I learn a ton looking through them :-)

**Comment by manofstick** [comment]
@KevinRansom ha! well that's good! if I do have any more ideas I'll give 'em a crack, but I think most of my ideas are a bit too radical to get slurped upstream...

**Comment by KevinRansom** [comment]
@manofstick , I am not so sure, I just think the rate of change of some stuff is so slow that FUD and real life(TM) gets in the way.

**Pull request closed without merging**